security-proxy/src/main/java/org/georchestra/security/LdapUserDetailsRequestHeaderProvider.java

```diff
@@ -39,163 +51,323 @@
-                    return Collections.emptyList();
-                }
-                headers = new ArrayList<Header>();
-                for (Map.Entry<String, String> entry : _headerMapping.entrySet()) {
-                    try {
-                        Attribute attributes = userData.getAttributes().get(entry.getValue());
-                        if (attributes != null) {
-                            NamingEnumeration<?> all = attributes.getAll();
-                            StringBuilder value = new StringBuilder();
-                            while (all.hasMore()) {
-                                if (value.length() > 0) {
-                                    value.append(',');
-                                }
-                                value.append(all.next());
-                            }
-                            headers.add(new BasicHeader(entry.getKey(), value.toString()));
-                        }
-                    } catch (javax.naming.NamingException e) {
-                        logger.error("problem adding headers for request:" + entry.getKey(), e);
+        headers.addAll(userDefinedHeaders);
+
+        return headers;
+    }
+
+    Map<String, String> getServiceMappings(String serviceName) {
+        Map<String, String> mappings = getDefaultMappings();
+        if (serviceName != null) {
+            Map<String, String> serviceMappings = this.perServiceMappings.get(serviceName);
+            if (serviceMappings != null) {
+                mappings = new HashMap<>(this.defaultMappinsg);
+                mappings.putAll(serviceMappings);
+            }
+        }
+        return mappings;
+    }
+
+    Map<String, String> getDefaultMappings() {
+        Map<String, String> mappings = this.defaultMappinsg;
+        return mappings;
+    }
+
+    private List<Header> buildStandardOrganizationHeaders(String username) {
+        // Add user organization
+        final String orgCn = loadOrgCn(username);
+
+        List<Header> headers = new ArrayList<>();
+        // add sec-orgname
+        if (orgCn != null) {
+            headers.add(new BasicHeader(SEC_ORG, orgCn));
+            try {
+                String dn = "cn=" + orgCn + "," + this.orgSearchBaseDN;
+                DirContextOperations ctx = this.ldapTemplate.lookupContext(dn);
+                headers.add(new BasicHeader(SEC_ORGNAME, ctx.getStringAttribute("o")));
+            } catch (RuntimeException ex) {
+                logger.warn("Cannot find associated org with cn " + orgCn);
+            }
+        }
+        return headers;
+    }
+
+    private String loadOrgCn(String username) {
+        try {
+            // Retreive memberOf attributes
+            // WARN! (groldan) looks like _userSearch is a singleton, so we could be mixing
+            // up setReturningAttributes from concurrent requests (we're synchronized on
+            // session here, not on _userSearch)
```

Comment:
for searching the LDAP, would it make more sense to create a new "userSearch" object each time instead of using a member object of the instance, and taking the risk of a race condition ?

Reply:
Fixed: 2e283fc9e5a94395359a31956b871cab3b6999d6